host/src/provider_db.rs

```diff
@@ -79,21 +88,80 @@ impl ProviderDb {
         offset: usize,
         num_storage_proofs: usize,
     ) -> Result<HashMap<Address, EIP1186AccountProofResponse>, anyhow::Error> {
-        let mut storage_proofs = HashMap::new();
+        let mut storage_proofs: HashMap<Address, EIP1186AccountProofResponse> = HashMap::new();
         let mut idx = offset;
-        for (address, keys) in storage_keys {
+
+        let mut storage_keys = storage_keys.clone();
+
+        let batch_limit = 1000;
+        while !storage_keys.is_empty() {
             inplace_print(&format!(
                 "fetching storage proof {idx}/{num_storage_proofs}..."
             ));
 
-            let indices = keys.iter().map(|x| x.to_be_bytes().into()).collect();
-            let proof = self.async_executor.block_on(async {
-                self.provider
-                    .get_proof(address, indices, Some(BlockId::from(block_number)))
-                    .await
-            })?;
-            storage_proofs.insert(address, proof);
-            idx += keys.len();
+            // Create a batch for all storage proofs
+            let mut batch = self.client.new_batch();
+
+            // Collect all requests
+            let mut requests = Vec::new();
+
+            let mut batch_size = 0;
+            while !storage_keys.is_empty() && batch_size < batch_limit {
+                let mut address_to_remove = None;
+                if let Some((address, keys)) = storage_keys.iter_mut().next() {
```

Comment:
https://doc.rust-lang.org/1.77.2/std/collections/struct.HashMap.html#method.retain
Using `retain` can be more simple

Reply:
It's challenging to use `retain` in this case. It would work nice if accounts can be removed when `keys.len() == 0`. However, we cannot remove all those accounts because even an account without any keys still needs the `get_proof` to get the account Merkle proof. So the only thing we can remove is this very specific account that was just processed.